pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/internal/JavaRuleUtil.java

```diff
@@ -513,4 +524,55 @@ public static boolean isDefaultValue(JTypeMirror type, ASTExpression expr) {
             return expr instanceof ASTNullLiteral;
         }
     }
+
+    /**
+     * Returns true if the expression is a {@link ASTNamedReferenceExpr}
+     * that references the symbol.
+     */
+    public static boolean isReferenceToVar(@Nullable ASTExpression expression, @NonNull JVariableSymbol symbol) {
+        if (expression instanceof ASTNamedReferenceExpr) {
+            return symbol.equals(((ASTNamedReferenceExpr) expression).getReferencedSym());
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if both expressions refer to the same variable.
+     * A "variable" here can also means a field path, eg, {@code this.field.a}.
+     * This method unifies {@code this.field} and {@code field} if possible,
+     * and also considers {@code this}.
+     *
+     * <p>Note that while this is more useful than just checking whether
+     * both expressions access the same symbol, it still does not mean that
+     * they both access the same <i>value</i>. The actual value is data-flow
+     * dependent.
+     */
+    public static boolean isReferenceToSameVar(ASTExpression e1, ASTExpression e2) {
+        if (e1 instanceof ASTNamedReferenceExpr && e2 instanceof ASTNamedReferenceExpr) {
+            if (!Objects.equals(((ASTNamedReferenceExpr) e2).getReferencedSym(),
+                                ((ASTNamedReferenceExpr) e1).getReferencedSym())) {
+                return false;
+            }
+
+            if (e1.getClass() != e2.getClass()) {
+                // unify `this.f` and `f`
+                // note, we already know that the symbol is the same so there's no scoping problem
+                return isSyntacticThisFieldAccess(e1) || isSyntacticThisFieldAccess(e2);
+            } else if (e1 instanceof ASTFieldAccess && e2 instanceof ASTFieldAccess) {
+                return isReferenceToSameVar(((ASTFieldAccess) e1).getQualifier(),
+                                            ((ASTFieldAccess) e2).getQualifier());
+            }
+            return e1 instanceof ASTVariableAccess && e2 instanceof ASTVariableAccess;
+        } else if (e1 instanceof ASTThisExpression || e2 instanceof ASTThisExpression) {
+            return e1.getClass() == e2.getClass();
```

Comment:
Almost wanted to say, we could just return "true" here -- but now I see, we are in an or-condition. So right, it's only the same, if both are a ASTThisExpression.